# Day 3: 2026-07-08

## Phase(s)
P0: Setup & Tooling | P1: Math Foundations

## Lessons Covered - Only Math - Linear Algebra
01 Linear Algebra Intuition; 02 Vectors, Matrices & Operations; 03 Matrix Transformations & Eigenvalues

## Est. Study Time (hrs)
6

## Resources (links)
https://aiengineeringfromscratch.com/catalog.html

## Daily Task - Vectors From Scratch
Read/watch the lesson material, then implement the from-scratch exercise in code (Python primary; TS/Rust/Julia optional per lesson). Run any provided tests before moving on.

## Assignment
Write a 5-6 bullet recap in your learning journal (what you built, one thing that was hard, one thing to revisit). Commit code + notes to your learning repo.

## Cumulative Project
Setup

## Project Application Task
Reproduce today's math concept in a Jupyter notebook using a CartQ dataset (e.g. order counts, prep times) instead of toy data.

### Dot product measures how much one vector points in same direction as other.


## TO DO
1. Matrix Matmul Implementation \/
2. BASIS AND RANK - understanding
3. Gram-Schmidt Process - understanding
4. Linear independence and projection from scratch (Python) - implemenation And Understanding
5. Rank, Projection, and QR with NumPy - Understanding
6. PyTorch -- Tensors Are Vectors with Autodiff - implementation
7. eigenvalues and eigenvectors and 03 কিচ্ছু বুজিনয় :) 
8. 01 exercise

## 01 Linear Algebra Intuition

In [1]:
print("Hello World")

Hello World


In [2]:
print("Mathar vitore gaja ghure gajar vitore matha")

Mathar vitore gaja ghure gajar vitore matha


In [3]:
import numpy as np

In [5]:
import numpy as np

In [6]:
import sys

print(sys.executable)
print(sys.version)

/Users/absyd/mystfs/cs/ai/coursingggs_/AI_ENG_FROM_SCRATCH/.venv/bin/python
3.12.13 (main, Mar 20 2026, 00:40:52) [Clang 22.1.1 ]


In [7]:
!source .venv/bin/activate

zsh:source:1: no such file or directory: .venv/bin/activate


In [8]:
!cd ../ && source .venv/bin/activate

In [9]:
import sys

print(sys.executable)
print(sys.version)

/Users/absyd/mystfs/cs/ai/coursingggs_/AI_ENG_FROM_SCRATCH/.venv/bin/python
3.12.13 (main, Mar 20 2026, 00:40:52) [Clang 22.1.1 ]


**Step 1: Vectors from scratch (Python)**

In [10]:
import math

In [11]:
class Vector:
    def __init__(self,components):
        self.components=list(components)
        self.dim=len(components)

    def __add__(self,other):
        # new_vector=[]
        # for i in range(self.components):
        #     new_vector.append(self.components[i]+other.components[i])

        # return Vector(new_vector)
        return Vector([a+b for a,b in zip(self.components,other.components)])

    def __sub__(self,other):
        # new_vector=[]
        # for i in range(self.components):
        #     new_vector.append(self.components[i]-other.components[i])

        # return Vector(new_vector)
        return Vector([a-b for a,b in zip(self.components,other.components)])

    def dot(self,other):
        # sum=0
        # for i in range(self.components):
        #     sum+=self.components[i]*other.components[i]
        # return sum
        return sum(a*b for a,b in zip(self.components,other.components))
                   
    def magnitude(self):
        # mag_sqaured=0
        # for i in range(self.components):
        #     mag_sqaured+=(self.components[i])**2
        # return math.sqrt(mag_sqaured)
        return sum(x**2 for x in self.components)**0.5

    def normalize(self):
        mag=self.magnitude()
        # normalize_arr=[]

        # for i in range(self.components):
        #     normalize_arr.append(self.components[i]/mag)
        # return Vector(normalize_arr)
        
        return Vector([x/mag for x in self.components])

    def cosine_similarity(self,other):
        # dot_product=self.dot(other)
        # similarity_value=dot_product/(self.magnitude()*other.magnitude())

        # return similarity_value
        return (self.dot(other))/(self.magnitude()*other.magnitude())

    def __repr__(self):
        return f"Vector({self.components})"
    


In [12]:

v1 = Vector([1, 2, 3])
v2 = Vector([4, 5, 6])

print(v1)
print(v2)

print("Addition:", v1 + v2)
print("Subtraction:", v1 - v2)
print("Dot Product:", v1.dot(v2))
print("Magnitude:", v1.magnitude())
print("Normalized:", v1.normalize())
print("Cosine Similarity:", v1.cosine_similarity(v2))

Vector([1, 2, 3])
Vector([4, 5, 6])
Addition: Vector([5, 7, 9])
Subtraction: Vector([-3, -3, -3])
Dot Product: 32
Magnitude: 3.7416573867739413
Normalized: Vector([0.2672612419124244, 0.5345224838248488, 0.8017837257372732])
Cosine Similarity: 0.9746318461970762


In [13]:
import math


class MyVector:

    def __init__(self, components):
        self.components = list(components)

        # Edge case 1: Empty vector
        if len(self.components) == 0:
            raise ValueError("Vector cannot be empty")

        # Edge case 2: Non-numeric values
        for value in self.components:
            if not isinstance(value, (int, float)):
                raise TypeError("All components must be numeric")

    @property
    def dim(self):
        return len(self.components)

    def _check_vector(self, other):
        # Edge case 3: Wrong operand type
        if not isinstance(other, MyVector):
            raise TypeError("Operand must be MyVector")

        # Edge case 4: Different dimensions
        if self.dim != other.dim:
            raise ValueError("Vectors must have same dimension")

    def __add__(self, other):
        self._check_vector(other)

        # Alternative definition
        # result = []
        # for i in range(self.dim):
        #     result.append(self.components[i] + other.components[i])
        # return MyVector(result)

        return MyVector([
            a + b for a, b in zip(self.components, other.components)
        ])

    def __sub__(self, other):
        self._check_vector(other)

        # Alternative definition
        # result = []
        # for i in range(self.dim):
        #     result.append(self.components[i] - other.components[i])
        # return MyVector(result)

        return MyVector([
            a - b for a, b in zip(self.components, other.components)
        ])

    def dot(self, other):
        self._check_vector(other)

        # Alternative definition
        # total = 0
        # for i in range(self.dim):
        #     total += self.components[i] * other.components[i]
        # return total

        return sum(
            a * b for a, b in zip(self.components, other.components)
        )

    def magnitude(self):

        # Alternative definition
        # mag_squared = 0
        # for value in self.components:
        #     mag_squared += value ** 2
        # return math.sqrt(mag_squared)

        return math.sqrt(
            sum(x ** 2 for x in self.components)
        )

    def normalize(self):
        mag = self.magnitude()

        # Edge case 5: Zero vector
        if mag == 0:
            raise ValueError("Cannot normalize zero vector")

        # Alternative definition
        # normalized = []
        # for value in self.components:
        #     normalized.append(value / mag)
        # return MyVector(normalized)

        return MyVector([
            x / mag for x in self.components
        ])

    def cosine_similarity(self, other):
        self._check_vector(other)

        mag1 = self.magnitude()
        mag2 = other.magnitude()

        # Edge case 6: Zero vector
        if mag1 == 0 or mag2 == 0:
            raise ValueError(
                "Cosine similarity undefined for zero vector"
            )

        # Alternative definition
        # dot_product = self.dot(other)
        # similarity = dot_product / (mag1 * mag2)
        # return similarity

        return self.dot(other) / (mag1 * mag2)

    def __repr__(self):
        return f"MyVector({self.components})"





In [14]:
# -------------------------
# Normal Usage
# -------------------------

v1 = MyVector([1, 2, 3])
v2 = MyVector([4, 5, 6])

print(v1)
print(v2)

print("Addition:", v1 + v2)
print("Subtraction:", v1 - v2)
print("Dot Product:", v1.dot(v2))
print("Magnitude:", v1.magnitude())
print("Normalized:", v1.normalize())
print("Cosine Similarity:", v1.cosine_similarity(v2))

MyVector([1, 2, 3])
MyVector([4, 5, 6])
Addition: MyVector([5, 7, 9])
Subtraction: MyVector([-3, -3, -3])
Dot Product: 32
Magnitude: 3.7416573867739413
Normalized: MyVector([0.2672612419124244, 0.5345224838248488, 0.8017837257372732])
Cosine Similarity: 0.9746318461970762


In [15]:


# -------------------------
# Edge Cases
# -------------------------

# Empty vector
try:
    MyVector([])
except Exception as e:
    print("Empty Vector:", e)

# Non-numeric values
try:
    MyVector([1, "hello", 3])
except Exception as e:
    print("Non-numeric:", e)

# Different dimensions
try:
    MyVector([1, 2]) + MyVector([1, 2, 3])
except Exception as e:
    print("Dimension Mismatch:", e)

# Wrong operand type
try:
    MyVector([1, 2]) + [3, 4]
except Exception as e:
    print("Wrong Operand:", e)

# Normalize zero vector
try:
    MyVector([0, 0, 0]).normalize()
except Exception as e:
    print("Normalize Zero Vector:", e)

# Cosine similarity with zero vector
try:
    MyVector([1, 2]).cosine_similarity(MyVector([0, 0]))
except Exception as e:
    print("Cosine Similarity Error:", e)

Empty Vector: Vector cannot be empty
Non-numeric: All components must be numeric
Dimension Mismatch: Vectors must have same dimension
Wrong Operand: Operand must be MyVector
Normalize Zero Vector: Cannot normalize zero vector
Cosine Similarity Error: Cosine similarity undefined for zero vector


### Step 2: Matrices from scratch (Python)

In [ ]:
# # Self Implementation 

# class Matrix:
#     def __init__(self,rows):
#         self.rows=[list(row) for row in rows]
#         self.shape=(len(rows),len(rows[0]))

#         for row in rows:
#             if len(row) != len(rows[0])):
#                 raise ValueError("All Of The Rows Must Be Same Dimenssion")

#     def __matmul__(self,other):
#         mat_2d=[]
        
        


    

In [17]:
class Matrix:
    def __init__(self, rows):
        self.rows = [list(row) for row in rows]
        self.shape = (len(self.rows), len(self.rows[0]))

    def __matmul__(self, other):
        if isinstance(other, Vector):
            return Vector([
                sum(self.rows[i][j] * other.components[j] for j in range(self.shape[1]))
                for i in range(self.shape[0])
            ])
        rows = []
        for i in range(self.shape[0]):
            row = []
            for j in range(other.shape[1]):
                row.append(sum(
                    self.rows[i][k] * other.rows[k][j]
                    for k in range(self.shape[1])
                ))
            rows.append(row)
        return Matrix(rows)

    def transpose(self):
        return Matrix([
            [self.rows[j][i] for j in range(self.shape[0])]
            for i in range(self.shape[1])
        ])

    def __repr__(self):
        return f"Matrix({self.rows})"


rotation_90 = Matrix([[0, -1], [1, 0]])
point = Vector([3, 1])
print("rotation_90",rotation_90)
rotated = rotation_90 @ point
print(f"Original: {point}")
print(f"Rotated 90°: {rotated}")

rotation_90 Matrix([[0, -1], [1, 0]])
Original: Vector([3, 1])
Rotated 90°: Vector([-1, 3])


In [18]:
import random

random.seed(42)
weights = Matrix([[random.gauss(0, 0.1) for _ in range(3)] for _ in range(2)])
input_vector = Vector([1.0, 0.5, -0.3])

output = weights @ input_vector
print(f"Input (3D): {input_vector}")
print(f"Output (2D): {output}")
print("This is what a neural network layer does -- matrix multiplication.")

Input (3D): Vector([1.0, 0.5, -0.3])
Output (2D): Vector([-0.019714737127338927, 0.10873956075097067])
This is what a neural network layer does -- matrix multiplication.


In [19]:
# Linear independence

def is_linearly_independent(vectors):
    n = len(vectors)
    dim = len(vectors[0].components)
    mat = Matrix([v.components[:] for v in vectors])
    rows = [row[:] for row in mat.rows]
    rank = 0
    for col in range(dim):
        pivot = None
        for row in range(rank, len(rows)):
            if abs(rows[row][col]) > 1e-10:
                pivot = row
                break
        if pivot is None:
            continue
        rows[rank], rows[pivot] = rows[pivot], rows[rank]
        scale = rows[rank][col]
        rows[rank] = [x / scale for x in rows[rank]]
        for row in range(len(rows)):
            if row != rank and abs(rows[row][col]) > 1e-10:
                factor = rows[row][col]
                rows[row] = [rows[row][j] - factor * rows[rank][j] for j in range(dim)]
        rank += 1
    return rank == n


def project(a, b):
    scalar = a.dot(b) / b.dot(b)
    return Vector([scalar * x for x in b.components])


def gram_schmidt(vectors):
    orthonormal = []
    for v in vectors:
        w = v
        for u in orthonormal:
            proj = project(w, u)
            w = w - proj
        if w.magnitude() < 1e-10:
            continue
        orthonormal.append(w.normalize())
    return orthonormal


v1 = Vector([1, 0, 0])
v2 = Vector([1, 1, 0])
v3 = Vector([1, 1, 1])
basis = gram_schmidt([v1, v2, v3])
for i, u in enumerate(basis):
    print(f"u{i+1} = {u}")
    print(f"  |u{i+1}| = {u.magnitude():.6f}")

print(f"u1 · u2 = {basis[0].dot(basis[1]):.6f}")
print(f"u1 · u3 = {basis[0].dot(basis[2]):.6f}")
print(f"u2 · u3 = {basis[1].dot(basis[2]):.6f}")

u1 = Vector([1.0, 0.0, 0.0])
  |u1| = 1.000000
u2 = Vector([0.0, 1.0, 0.0])
  |u2| = 1.000000
u3 = Vector([0.0, 0.0, 1.0])
  |u3| = 1.000000
u1 · u2 = 0.000000
u1 · u3 = 0.000000
u2 · u3 = 0.000000


In [20]:
import random

random.seed(42)
weights = Matrix([[random.gauss(0, 0.1) for _ in range(3)] for _ in range(2)])
print("Weights : ",weights)

input_vector = Vector([1.0, 0.5, -0.3])
print("input_vector : ",input_vector)

output = weights @ input_vector

print(f"Input (3D): {input_vector}")
print(f"Output (2D): {output}")
print("This is what a neural network layer does -- matrix multiplication.")

Weights :  Matrix([[-0.014409032957792837, -0.01729036003315193, -0.011131586156766248], [0.07019837250988631, -0.01275882837828871, -0.14973534143409575]])
input_vector :  Vector([1.0, 0.5, -0.3])
Input (3D): Vector([1.0, 0.5, -0.3])
Output (2D): Vector([-0.019714737127338927, 0.10873956075097067])
This is what a neural network layer does -- matrix multiplication.


In [21]:
print([[random.gauss(0, 0.1) for _ in range(3)] for _ in range(2)])

[[0.03323183440677153, -0.026733747849716823, -0.021695868414519502], [0.011588478670085507, 0.023229773690672088, 0.1163558686599143]]


### Now the same thing with NumPy -- what you'll actually use in practice:


In [22]:
import numpy as np

a = np.array([1,2,3],dtype=float)
b = np.array([4,5,6],dtype=float)

In [24]:
print(f"a + b = {a + b}")
print(f"a · b = {np.dot(a,b)}")
print(f"|a| = {np.linalg.norm(a)}")
print(f"cosine = {np.dot(a, b) / (np.linalg.norm(a) * np.linalg.norm(b)):.4f}")


a + b = [5. 7. 9.]
a · b = 32.0
|a| = 3.7416573867739413
cosine = 0.9746


### Rank, Projection, and QR with NumPy

In [26]:
import numpy as np
A = np.array([[1,2],[3,4]])


In [28]:
print(f"Rank : {np.linalg.matrix_rank(A)}")

Rank : 2


In [29]:
a = np.array([3, 4])
b = np.array([1, 0])

In [30]:
# Projection
proj = (np.dot(a, b) / np.dot(b, b)) * b
print(f"Projection of {a} onto {b}: {proj}")

Projection of [3 4] onto [1 0]: [3. 0.]


In [31]:
Q, R = np.linalg.qr(np.random.randn(3, 3))
print(f"Q is orthogonal: {np.allclose(Q @ Q.T, np.eye(3))}")
print(f"R is upper triangular: {np.allclose(R, np.triu(R))}")


Q is orthogonal: True
R is upper triangular: True


In [32]:
import torch

x = torch.randn(3, requires_grad=True)
y = torch.tensor([1.0, 0.0, 0.0])

similarity = torch.dot(x, y)
similarity.backward()

print(f"x = {x.data}")
print(f"y = {y.data}")
print(f"dot product = {similarity.item():.4f}")
print(f"d(dot)/dx = {x.grad}")

ModuleNotFoundError: No module named 'torch'

**Exercises:**
1. Implement Vector.angle_between(other) that returns the angle in degrees between two vectors
Create a 2D scaling matrix that doubles the x-coordinate and triples the y-coordinate, then apply it to the vector [1, 1]
2. Given 5 random word-like vectors (dimension 50), find the two most similar using cosine similarity
3.Verify that the Gram-Schmidt output is truly orthonormal: check that every pair has dot product 0 and every vector has magnitude 1
4. Create a 3x3 matrix with rank 2. Verify using the rank() method. Then explain what geometric object the columns span.
5. Project the vector [1, 2, 3] onto [1, 1, 1]. What does the result represent geometrically?


## 02 Vectors, Matrices & Operations


In [8]:
class Vector:
    def __init__(self,components):
        self.components=list(components)
        self.dim=len(components)

    def __add__(self,other):
        # new_vector=[]
        # for i in range(self.components):
        #     new_vector.append(self.components[i]+other.components[i])

        # return Vector(new_vector)
        return Vector([a+b for a,b in zip(self.components,other.components)])

    def __sub__(self,other):
        # new_vector=[]
        # for i in range(self.components):
        #     new_vector.append(self.components[i]-other.components[i])

        # return Vector(new_vector)
        return Vector([a-b for a,b in zip(self.components,other.components)])

    def dot(self,other):
        # sum=0
        # for i in range(self.components):
        #     sum+=self.components[i]*other.components[i]
        # return sum
        return sum(a*b for a,b in zip(self.components,other.components))
                   
    def magnitude(self):
        # mag_sqaured=0
        # for i in range(self.components):
        #     mag_sqaured+=(self.components[i])**2
        # return math.sqrt(mag_sqaured)
        return sum(x**2 for x in self.components)**0.5

    def normalize(self):
        mag=self.magnitude()
        # normalize_arr=[]

        # for i in range(self.components):
        #     normalize_arr.append(self.components[i]/mag)
        # return Vector(normalize_arr)
        
        return Vector([x/mag for x in self.components])

    def cosine_similarity(self,other):
        # dot_product=self.dot(other)
        # similarity_value=dot_product/(self.magnitude()*other.magnitude())

        # return similarity_value
        return (self.dot(other))/(self.magnitude()*other.magnitude())

    def __repr__(self):
        return f"Vector({self.components})"
    


In [13]:
import math


class Matrix:

    def __init__(self, rows):

        if not rows:
            raise ValueError("Matrix cannot be empty")

        self.rows = [list(row) for row in rows]

        cols = len(self.rows[0])

        for row in self.rows:
            if len(row) != cols:
                raise ValueError("All rows must have same length")

        self.shape = (len(self.rows), cols)

    def __add__(self, other):

        if not isinstance(other, Matrix):
            return NotImplemented

        if self.shape != other.shape:
            raise ValueError("Matrices must have same shape")

        result = []

        for i in range(self.shape[0]):

            row = []

            for j in range(self.shape[1]):
                row.append(self.rows[i][j] + other.rows[i][j])

            result.append(row)

        return Matrix(result)

    def __sub__(self, other):

        if not isinstance(other, Matrix):
            return NotImplemented

        if self.shape != other.shape:
            raise ValueError("Matrices must have same shape")

        result = []

        for i in range(self.shape[0]):

            row = []

            for j in range(self.shape[1]):
                row.append(self.rows[i][j] - other.rows[i][j])

            result.append(row)

        return Matrix(result)

    def scalar_multiply(self, number):

        if not isinstance(number, (int, float)):
            return NotImplemented

        result = []

        for row in self.rows:
            result.append([number * x for x in row])

        return Matrix(result)

    def element_wise_multiply(self, other):

        if not isinstance(other, Matrix):
            return NotImplemented

        if self.shape != other.shape:
            raise ValueError("Matrices must have same shape")

        result = []

        for i in range(self.shape[0]):

            row = []

            for j in range(self.shape[1]):
                row.append(self.rows[i][j] * other.rows[i][j])

            result.append(row)

        return Matrix(result)

    def __matmul__(self, other):

        if isinstance(other, Vector):

            if self.shape[1] != other.dim:
                raise ValueError("Incompatible dimensions")

            return Vector([
                sum(
                    self.rows[i][j] * other.components[j]
                    for j in range(self.shape[1])
                )
                for i in range(self.shape[0])
            ])

        if not isinstance(other, Matrix):
            return NotImplemented

        if self.shape[1] != other.shape[0]:
            raise ValueError("Incompatible matrix dimensions")

        result = []

        for i in range(self.shape[0]):

            row = []

            for j in range(other.shape[1]):

                value = sum(
                    self.rows[i][k] * other.rows[k][j]
                    for k in range(self.shape[1])
                )

                row.append(value)

            result.append(row)

        return Matrix(result)

    def transpose(self):

        result = []

        for i in range(self.shape[1]):

            row = []

            for j in range(self.shape[0]):
                row.append(self.rows[j][i])

            result.append(row)

        return Matrix(result)

    def __repr__(self):

        rows = "\n  ".join(str(row) for row in self.rows)

        return f"Matrix({self.shape}):\n  {rows}"

In [14]:
m1 = Matrix([
    [1, 2],
    [3, 4]
])

m2 = Matrix([
    [5, 6],
    [7, 8]
])

print("m1")
print(m1)

print("\nm2")
print(m2)

print("\nAddition")
print(m1 + m2)

print("\nSubtraction")
print(m1 - m2)

print("\nScalar Multiply")
print(m1.scalar_multiply(3))

print("\nElement-wise Multiply")
print(m1.element_wise_multiply(m2))

print("\nTranspose")
print(m1.transpose())

print("\nMatrix Multiplication")
print(m1 @ m2)

v = Vector([1, 2])

print("\nMatrix × Vector")
print(m1 @ v)

m1
Matrix((2, 2)):
  [1, 2]
  [3, 4]

m2
Matrix((2, 2)):
  [5, 6]
  [7, 8]

Addition
Matrix((2, 2)):
  [6, 8]
  [10, 12]

Subtraction
Matrix((2, 2)):
  [-4, -4]
  [-4, -4]

Scalar Multiply
Matrix((2, 2)):
  [3, 6]
  [9, 12]

Element-wise Multiply
Matrix((2, 2)):
  [5, 12]
  [21, 32]

Transpose
Matrix((2, 2)):
  [1, 3]
  [2, 4]

Matrix Multiplication
Matrix((2, 2)):
  [19, 22]
  [43, 50]

Matrix × Vector
Vector([5, 11])


In [4]:

rotation_90 = Matrix([[0, -1], [1, 0]])
point = Vector([3, 1])
print("rotation_90",rotation_90)
rotated = rotation_90 @ point
print(f"Original: {point}")
print(f"Rotated 90°: {rotated}")

rotation_90 Matrix([[0, -1], [1, 0]])
Original: Vector([3, 1])
Rotated 90°: Vector([-1, 3])


``` output = activation(weights @ input + bias)```


In [5]:
v = [3, 4]        # -- a 2D vector
w = [1, 0, -2]    # -- a 3D vector

In neural networks, weight matrices transform input vectors into output vectors. A layer with 784 inputs and 128 outputs uses a 128x784 weight matrix.

Matrix multiplication has a strict rule: (m x n) @ (n x p) = (m x p). The inner dimensions must match.

```
(128 x 784) @ (784 x 1) = (128 x 1)
  weights       input       output

Inner dimensions: 784 = 784  -- valid
```

Operation	What it does	Neural network use
Addition	Element-wise combine	Adding bias to output
Scalar multiply	Scale every element	Learning rate * gradients
Matrix multiply	Transform vectors	Layer forward pass
Transpose	Flip rows and columns	Backpropagation
Determinant	Single number summary	Checking invertibility
Inverse	Undo a transformation	Solving linear systems
Identity	Do-nothing matrix	Initialization, residual connections

### Element-wise vs matrix multiplication

1. Element-wise: multiply matching positions. Both matrices must be the same shape.
```
| 1  2 |   | 5  6 |   | 5  12 |
| 3  4 | * | 7  8 | = | 21 32 |
```

2. Matrix multiplication: dot products of rows and columns. Inner dimensions must match.
```
| 1  2 |   | 5  6 |   | 1*5+2*7  1*6+2*8 |   | 19  22 |
| 3  4 | @ | 7  8 | = | 3*5+4*7  3*6+4*8 | = | 43  50 |
```

### Broadcasting
When you add a bias vector to a matrix of outputs, the shapes do not match. Broadcasting stretches the smaller array to fit.

```
| 1  2  3 |   +   [10, 20, 30]
| 4  5  6 |

Broadcasting stretches the vector across rows:

| 1  2  3 |   | 10  20  30 |   | 11  22  33 |
| 4  5  6 | + | 10  20  30 | = | 14  25  36 |
```


In [3]:
# latest matrix implementation
class Matrix:
    def __init__(self, data):
        self.data = [list(row) for row in data]
        self.rows = len(self.data)
        self.cols = len(self.data[0])
        self.shape = (self.rows, self.cols)

    def __repr__(self):
        rows_str = "\n  ".join(str(row) for row in self.data)
        return f"Matrix({self.shape}):\n  {rows_str}"

    def __add__(self, other):
        return Matrix([
            [self.data[i][j] + other.data[i][j] for j in range(self.cols)]
            for i in range(self.rows)
        ])

    def __sub__(self, other):
        return Matrix([
            [self.data[i][j] - other.data[i][j] for j in range(self.cols)]
            for i in range(self.rows)
        ])

    def scalar_multiply(self, scalar):
        return Matrix([
            [self.data[i][j] * scalar for j in range(self.cols)]
            for i in range(self.rows)
        ])

    def element_wise_multiply(self, other):
        return Matrix([
            [self.data[i][j] * other.data[i][j] for j in range(self.cols)]
            for i in range(self.rows)
        ])

    def matmul(self, other):
        return Matrix([
            [
                sum(self.data[i][k] * other.data[k][j] for k in range(self.cols))
                for j in range(other.cols)
            ]
            for i in range(self.rows)
        ])

    def transpose(self):
        return Matrix([
            [self.data[j][i] for j in range(self.rows)]
            for i in range(self.cols)
        ])

    def determinant(self):
        if self.shape == (1, 1):
            return self.data[0][0]
        if self.shape == (2, 2):
            return self.data[0][0] * self.data[1][1] - self.data[0][1] * self.data[1][0]
        det = 0
        for j in range(self.cols):
            minor = Matrix([
                [self.data[i][k] for k in range(self.cols) if k != j]
                for i in range(1, self.rows)
            ])
            det += ((-1) ** j) * self.data[0][j] * minor.determinant()
        return det

    def inverse_2x2(self):
        det = self.determinant()
        if det == 0:
            raise ValueError("Matrix is singular, no inverse exists")
        return Matrix([
            [self.data[1][1] / det, -self.data[0][1] / det],
            [-self.data[1][0] / det, self.data[0][0] / det]
        ])


    @staticmethod
    def identity(n):
        return Matrix([
            [1 if i == j else 0 for j in range(n)]
            for i in range(n)
        ])

In [4]:
A = Matrix([[1, 2], [3, 4]])
B = Matrix([[5, 6], [7, 8]])

print("A + B =", (A + B).data)
print("A @ B =", A.matmul(B).data)
print("A^T =", A.transpose().data)
print("det(A) =", A.determinant())
print("A^-1 =", A.inverse_2x2().data)

I = Matrix.identity(2)
print("A @ A^-1 =", A.matmul(A.inverse_2x2()).data)

A + B = [[6, 8], [10, 12]]
A @ B = [[19, 22], [43, 50]]
A^T = [[1, 3], [2, 4]]
det(A) = -2
A^-1 = [[-2.0, 1.0], [1.5, -0.5]]
A @ A^-1 = [[1.0, 0.0], [0.0, 1.0]]


In [17]:
#  Implement on Neural Networks

import random

In [20]:
inputs = Matrix([[0.5],[0.8],[0.2]])
inputs

Matrix((3, 1)):
  [0.5]
  [0.8]
  [0.2]

In [19]:
weights = Matrix(
    [
        [random.uniform(-1,1) for _ in range(3)] for _ in range(2)
    ]
)

weights

Matrix((2, 3)):
  [0.029522615615595438, -0.9796373663405804, -0.33807522400082624]
  [0.8189002641174659, -0.9533173124591616, -0.024085980645941696]

In [23]:
bias = Matrix([[0.1], [0.1]])

In [21]:
def relu_matrix(m):
    return Matrix([[max(0,val) for val in row] for row in m.data])

In [22]:
relu_matrix(weights)

Matrix((2, 3)):
  [0.029522615615595438, 0, 0]
  [0.8189002641174659, 0, 0]

In [24]:
pre_activation = weights.matmul(inputs)+bias

In [25]:
pre_activation

Matrix((2, 1)):
  [-0.7365636300648319]
  [-0.2580209140377847]

In [26]:
output=relu_matrix(pre_activation)
output

Matrix((2, 1)):
  [0]
  [0]

In [27]:
import random

inputs = Matrix([[0.5], [0.8], [0.2]])
weights = Matrix([
    [random.uniform(-1, 1) for _ in range(3)]
    for _ in range(2)
])
bias = Matrix([[0.1], [0.1]])

def relu_matrix(m):
    return Matrix([[max(0, val) for val in row] for row in m.data])

pre_activation = weights.matmul(inputs) + bias
output = relu_matrix(pre_activation)

print(f"Input shape: {inputs.shape}")
print(f"Weight shape: {weights.shape}")
print(f"Output shape: {output.shape}")
print(f"Output: {output.data}")

Input shape: (3, 1)
Weight shape: (2, 3)
Output shape: (2, 1)
Output: [[0], [0]]


In [29]:
import numpy as np
import numpy as np

A = np.array([[1, 2], [3, 4]])
B = np.array([[5, 6], [7, 8]])

print("A",A)
print("B",B)
print("A + B =\n", A + B)
print("A * B (element-wise) =\n", A * B)
print("A @ B (matrix multiply) =\n", A @ B)
print("A^T =\n", A.T)
print("det(A) =", np.linalg.det(A))
print("A^-1 =\n", np.linalg.inv(A))
print("I =\n", np.eye(2))

inputs = np.random.randn(3, 1)
weights = np.random.randn(2, 3)
bias = np.array([[0.1], [0.1]])
output = np.maximum(0, weights @ inputs + bias)

print(f"\nNeural network layer: {weights.shape} @ {inputs.shape} = {output.shape}")
print(f"Output:\n{output}")

A [[1 2]
 [3 4]]
B [[5 6]
 [7 8]]
A + B =
 [[ 6  8]
 [10 12]]
A * B (element-wise) =
 [[ 5 12]
 [21 32]]
A @ B (matrix multiply) =
 [[19 22]
 [43 50]]
A^T =
 [[1 3]
 [2 4]]
det(A) = -2.0000000000000004
A^-1 =
 [[-2.   1. ]
 [ 1.5 -0.5]]
I =
 [[1. 0.]
 [0. 1.]]

Neural network layer: (2, 3) @ (3, 1) = (2, 1)
Output:
[[0.35958401]
 [0.02636437]]


In [31]:
matrix = np.array([[1,2,3],[1,2,3]])
matrix

array([[1, 2, 3],
       [1, 2, 3]])

In [32]:
bias = np.array([10,20,30])
bias

array([10, 20, 30])

In [33]:
print(matrix+bias)

[[11 22 33]
 [11 22 33]]


### Exercises

1. Verify the inverse. Multiply A @ A.inverse_2x2() and confirm you get the identity matrix. Try it with three different 2x2 matrices. What happens when the determinant is zero? <span style="background-color:green;color:white;font-weight:bold;padding:3px 10px;border-radius:20px;">DONE</span>
2. Implement 3x3 inverse. Extend the Matrix class to compute inverses for 3x3 matrices using the adjugate method. Test it against NumPy's np.linalg.inv.
3. Build a two-layer network. Using only your Matrix class (no NumPy), create a two-layer neural network: input (3) -> hidden (4) -> output (2). Initialize random weights, run a forward pass, and verify all shapes are correct. <span style="background-color:green;color:white;font-weight:bold;padding:3px 10px;border-radius:20px;">DONE</span>



In [38]:
mat_a=Matrix([[1,2],[3,4]])

In [40]:
multiply_val=mat_a.matmul(mat_a)
multiply_val

Matrix((2, 2)):
  [7, 10]
  [15, 22]

In [41]:
multiply_val.inverse_2x2()

Matrix((2, 2)):
  [5.5, -2.5]
  [-3.75, 1.75]

In [5]:
# 3. based on the first one
import random

inputs = Matrix([[0.5], [0.8], [0.2]])
weights = Matrix([
    [random.uniform(-1, 1) for _ in range(3)]
    for _ in range(2)
])
bias = Matrix([[0.1], [0.1]])

def relu_matrix(m):
    return Matrix([[max(0, val) for val in row] for row in m.data])

pre_activation = weights.matmul(inputs) + bias
output = relu_matrix(pre_activation)



print(f"Input shape: {inputs.shape}")
print(f"Weight shape: {weights.shape}")
print(f"Output shape: {output.shape}")
print(f"Output: {output.data}")

Input shape: (3, 1)
Weight shape: (2, 3)
Output shape: (2, 1)
Output: [[0], [0]]


In [44]:
# output will be feed as next layer input 
# Input shape: (2, 1)
# 

weights_02 = Matrix([
    [random.uniform(-1, 1) for _ in range(2)]
    for _ in range(1)
])
bias_02 = Matrix([[0.3]])

output_02= weights_02.matmul(output)+bias
print(f"Input shape: {output.shape}")
print(f"Weight 02 shape: {weights_02.shape}")
print(f"Output 02 shape: {output_02.shape}")
print(f"Output 02: {output.data}")

Input shape: (2, 1)
Weight 02 shape: (1, 2)
Output 02 shape: (1, 1)
Output 02: [[0.22971424781632752], [0.007423426153830265]]


In [ ]:
# 3. 
import random

inputs = Matrix([[0.5], [0.8], [0.2]])
weights = Matrix([
    [random.uniform(-1, 1) for _ in range(3)]
    for _ in range(2)
])
bias = Matrix([[0.1], [0.1]])

def relu_matrix(m):
    return Matrix([[max(0, val) for val in row] for row in m.data])

pre_activation = weights.matmul(inputs) + bias
output = relu_matrix(pre_activation)



print(f"Input shape: {inputs.shape}")
print(f"Weight shape: {weights.shape}")
print(f"Output shape: {output.shape}")
print(f"Output: {output.data}")

# output will be feed as next layer input 
# Input shape: (2, 1)
# 

weights_02 = Matrix([
    [random.uniform(-1, 1) for _ in range(2)]
    for _ in range(1)
])
bias_02 = Matrix([[0.3]])

output_02= weights_02.matmul(output)+bias
print(f"Input shape: {output.shape}")
print(f"Weight 02 shape: {weights_02.shape}")
print(f"Output 02 shape: {output_02.shape}")
print(f"Output 02: {output.data}")

In [ ]:
inputs = Matrix([
    [0.5],
    [0.8],
    [0.2]
])                     # (3,1)

weights1 = Matrix([
    [random.uniform(-1,1) for _ in range(3)]
    for _ in range(4)
])                     # (4,3)

bias1 = Matrix([
    [0.1]
    for _ in range(4)
])                     # (4,1)

hidden = relu_matrix((weights1 @ inputs) + bias1)

weights2 = Matrix([
    [random.uniform(-1,1) for _ in range(4)]
    for _ in range(2)
])                     # (2,4)

bias2 = Matrix([
    [0.2]
    for _ in range(2)
])                     # (2,1)

output = (weights2 @ hidden) + bias2


print(inputs.shape)
print(weights1.shape)
print(hidden.shape)
print(weights2.shape)
print(output.shape)

#### 03 

Build a two-layer network. Using only your Matrix class (no NumPy), create a two-layer neural network: input (3) -> hidden (4) -> output (2). Initialize random weights, run a forward pass, and verify all shapes are correct.


In [7]:
# latest matrix implementation
class Matrix:
    def __init__(self, data):
        self.data = [list(row) for row in data]
        self.rows = len(self.data)
        self.cols = len(self.data[0])
        self.shape = (self.rows, self.cols)

    def __repr__(self):
        rows_str = "\n  ".join(str(row) for row in self.data)
        return f"Matrix({self.shape}):\n  {rows_str}"

    def __add__(self, other):
        return Matrix([
            [self.data[i][j] + other.data[i][j] for j in range(self.cols)]
            for i in range(self.rows)
        ])

    def __sub__(self, other):
        return Matrix([
            [self.data[i][j] - other.data[i][j] for j in range(self.cols)]
            for i in range(self.rows)
        ])

    def scalar_multiply(self, scalar):
        return Matrix([
            [self.data[i][j] * scalar for j in range(self.cols)]
            for i in range(self.rows)
        ])

    def element_wise_multiply(self, other):
        return Matrix([
            [self.data[i][j] * other.data[i][j] for j in range(self.cols)]
            for i in range(self.rows)
        ])

    def matmul(self, other):
        return Matrix([
            [
                sum(self.data[i][k] * other.data[k][j] for k in range(self.cols))
                for j in range(other.cols)
            ]
            for i in range(self.rows)
        ])

    def transpose(self):
        return Matrix([
            [self.data[j][i] for j in range(self.rows)]
            for i in range(self.cols)
        ])

    def determinant(self):
        if self.shape == (1, 1):
            return self.data[0][0]
        if self.shape == (2, 2):
            return self.data[0][0] * self.data[1][1] - self.data[0][1] * self.data[1][0]
        det = 0
        for j in range(self.cols):
            minor = Matrix([
                [self.data[i][k] for k in range(self.cols) if k != j]
                for i in range(1, self.rows)
            ])
            det += ((-1) ** j) * self.data[0][j] * minor.determinant()
        return det

    def inverse_2x2(self):
        det = self.determinant()
        if det == 0:
            raise ValueError("Matrix is singular, no inverse exists")
        return Matrix([
            [self.data[1][1] / det, -self.data[0][1] / det],
            [-self.data[1][0] / det, self.data[0][0] / det]
        ])


    @staticmethod
    def identity(n):
        return Matrix([
            [1 if i == j else 0 for j in range(n)]
            for i in range(n)
        ])

In [11]:
# my imlementation
import random
bias=Matrix([[0.1],[0.03],[0.04],[0.008]])

def relu_matrix(m):
    return Matrix([[max(0, val) for val in row] for row in m.data])

input_1 = Matrix([[1],[2],[3]]) #3x1

weights_1 = Matrix([
    [random.uniform(-1,1) for _ in range(3)]
    for _ in range(4)
])# 4x3  
pre_activation_1=weights_1.matmul(input_1)
output_1=relu_matrix(pre_activation_1)

input_2=output_1 #4x1
weights_2 = Matrix([
    [random.uniform(-1,1) for _ in range(4)]
    for _ in range(2)
])# 2x4   

pre_activation_2=weights_2.matmul(input_2)
output_2=relu_matrix(pre_activation_2)
output_2
# no need this as per the assignment
# input_3=output_2 #2x1
# weights_3 = Matrix([
#     [random.uniform(-1,1) for _ in range(2)]
#     for _ in range(2)
# ])# 2x2  
# pre_activation_3=weights_3.matmul(input_3)+bias
# output_3=relu_matrix(pre_activation_3)






### GOOD TO USE

In [34]:
import random

def relu_matrix(m):
    return Matrix([[max(0, val) for val in row] for row in m.data])


# -------------------------
# Input (3x1)
# -------------------------
inputs = Matrix([
    [1],
    [2],
    [3]
])

# -------------------------
# Layer 1
# -------------------------
weights_1 = Matrix([
    [random.uniform(-1, 1) for _ in range(3)]
    for _ in range(4)
])  # 4x3

bias_1 = Matrix([
    [0.1],
    [0.1],
    [0.1],
    [0.1]
])  # 4x1

pre_activation_1 = weights_1.matmul(inputs) + bias_1
hidden = relu_matrix(pre_activation_1)

# -------------------------
# Layer 2
# -------------------------
weights_2 = Matrix([
    [random.uniform(-1, 1) for _ in range(4)]
    for _ in range(2)
])  # 2x4

bias_2 = Matrix([
    [0.1],
    [0.1]
])  # 2x1

pre_activation_2 = weights_2.matmul(hidden) + bias_2
output = relu_matrix(pre_activation_2)

# -------------------------
# Print results
# -------------------------
print("Input:")
print(inputs)

print("\nWeights 1:")
print(weights_1)

print("\nBias 1:")
print(bias_1)

print("\nHidden pre-activation:")
print(pre_activation_1)

print("\nHidden output:")
print(hidden)

print("\nWeights 2:")
print(weights_2)

print("\nBias 2:")
print(bias_2)

print("\nOutput pre-activation:")
print(pre_activation_2)

print("\nFinal output:")
print(output)

# -------------------------
# Shape check
# -------------------------
print("\n=== Shape Check ===")
print("inputs:", inputs.shape)
print("weights_1:", weights_1.shape)
print("bias_1:", bias_1.shape)
print("hidden:", hidden.shape)
print("weights_2:", weights_2.shape)
print("bias_2:", bias_2.shape)
print("output:", output.shape)

Input:
Matrix((3, 1)):
  [1]
  [2]
  [3]

Weights 1:
Matrix((4, 3)):
  [-0.6716735332967259, 0.6558642731654154, -0.31507217084397854]
  [-0.29618708062047516, 0.2805547422662267, -0.9833286626225994]
  [0.8352644646792688, 0.7188775309615163, -0.14111147554981884]
  [0.23912187930110518, -0.45280341375473765, 0.8414698168127253]

Bias 1:
Matrix((4, 1)):
  [0.1]
  [0.1]
  [0.1]
  [0.1]

Hidden pre-activation:
Matrix((4, 1)):
  [-0.20516149949783066]
  [-2.58506358395582]
  [1.949685099952845]
  [1.957924502229806]

Hidden output:
Matrix((4, 1)):
  [0]
  [0]
  [1.949685099952845]
  [1.957924502229806]

Weights 2:
Matrix((2, 4)):
  [0.022817201900853235, 0.14650500027935065, 0.561099665753765, -0.46266357743632813]
  [0.0601880717372929, 0.0005374611831994169, 0.6253086364475984, -0.9038954283156428]

Bias 2:
Matrix((2, 1)):
  [0.1]
  [0.1]

Output pre-activation:
Matrix((2, 1)):
  [0.28810730335675316]
  [-0.45060407519898915]

Final output:
Matrix((2, 1)):
  [0.28810730335675316]
  [0]

In [12]:
weights_1 = Matrix([
    [random.uniform(-1,1) for _ in range(3)]
    for _ in range(4)
])# 4x3              
weights_1

Matrix((4, 3)):
  [-0.4236603212185359, 0.3478594422984804, -0.4487722199640367]
  [-0.5562210013072302, -0.2489018673186949, 0.9087693296228525]
  [0.1442550901071391, 0.4912419512623203, 0.3714838745440947]
  [-0.46996533296773224, 0.44766604363746176, 0.2348514050120447]

In [13]:
input_1

Matrix((3, 1)):
  [1]
  [2]
  [3]

In [17]:
bias=Matrix([[0.1],[0.03],[0.04],[0.008]])

In [18]:
def relu_matrix(m):
    return Matrix([[max(0, val) for val in row] for row in m.data])

In [20]:
pre_activation_1=weights_1.matmul(input_1)+bias
pre_activation_1

Matrix((4, 1)):
  [-0.9742580965136852]
  [1.7022832529239376]
  [2.281190616264064]
  [1.1379209693433254]

In [22]:
output_1=relu_matrix(pre_activation_1)
output_1

Matrix((4, 1)):
  [0]
  [1.7022832529239376]
  [2.281190616264064]
  [1.1379209693433254]

In [23]:
input_2=output_1 #4x1

In [24]:
weights_2 = Matrix([
    [random.uniform(-1,1) for _ in range(4)]
    for _ in range(2)
])# 2x4   
weights_2

Matrix((2, 4)):
  [-0.6418711382042652, -0.9027637839269564, -0.905360307654284, -0.7095686001087842]
  [0.6525287221000939, -0.9919964124360874, 0.38702810098505536, -0.34471541358477786]

In [26]:
pre_activation_2=weights_2.matmul(input_2)+bias
pre_activation_2

Matrix((2, 1)):
  [-4.309492098135375]
  [-1.1680329052269458]

In [28]:
output_2=relu_matrix(pre_activation_2)
output_2

Matrix((2, 1)):
  [0]
  [0]

In [29]:
input_3=output_2 #2x1
weights_3 = Matrix([
    [random.uniform(-1,1) for _ in range(2)]
    for _ in range(2)
])# 2x2  
pre_activation_3=weights_3.matmul(input_3)+bias
output_3=relu_matrix(pre_activation_3)


In [31]:
weights_3 = Matrix([
    [random.uniform(-1,1) for _ in range(2)]
    for _ in range(2)
])# 2x2   
weights_3

Matrix((2, 2)):
  [0.03797198763086951, -0.25597287218026765]
  [-0.5427364152698997, 0.6624455877199895]

In [32]:
pre_activation_3=weights_3.matmul(input_3)+bias
pre_activation_3

Matrix((2, 1)):
  [0.1]
  [0.03]

In [33]:
output_3=relu_matrix(pre_activation_3)
output_3

Matrix((2, 1)):
  [0.1]
  [0.03]

## 03 Matrix Transformations (need deepr understanding)

```
A @ v = lambda * v

v is the eigenvector (direction that survives)
lambda is the eigenvalue (how much it stretches)

Example: A = | 2  1 |
             | 1  2 |

Eigenvector [1, 1] with eigenvalue 3:
  A @ [1,1] = [3, 3] = 3 * [1, 1]     (same direction, scaled by 3)

Eigenvector [1, -1] with eigenvalue 1:
  A @ [1,-1] = [1, -1] = 1 * [1, -1]  (same direction, unchanged)

  A = V @ D @ V^(-1)

V = matrix whose columns are eigenvectors
D = diagonal matrix of eigenvalues
V^(-1) = inverse of V

This says: rotate into eigenvector coordinates, scale along each axis, rotate back.

  ```

**Exercises**
1. Apply rotation, scaling, and shearing to a unit square (corners at [0,0], [1,0], [1,1], [0,1]). Print the transformed corners for each. Verify that rotation preserves distances between corners.
2. Find the eigenvalues of the matrix [[4, 2], [1, 3]] by hand using the characteristic equation. Then verify with your from-scratch function and with NumPy.
3. Create a composition of three transformations (rotate 30 degrees, scale by [1.5, 0.8], shear with kx=0.3) and apply it to 8 points arranged in a circle. Print before and after coordinates. Compute the determinant of the composed matrix and verify it equals the product of the individual determinants.


Think about people.

Need 3 people move table.

### Case 1: Linearly Independent

Person A pushes east.

→

Person B pushes north.

↑

Person C pushes upward.

Out of screen.

All three do different jobs.

Remove any one person.

Cannot replace them using other two.

This is **linearly independent**.

They can reach any point in 3D space.

---

### Case 2: Linearly Dependent

Person A:

→

Person B:

↑

Person C:

↗

But

```
Person C = Person A + Person B
```

Third person adds nothing new.

Already can make C by combining A and B.

So only 2 independent directions exist.

Span becomes plane instead of whole 3D space.

---

Simple vector example

```
v1 = (1, 0)
v2 = (0, 1)
```

Independent.

Cannot make `(1,0)` from `(0,1)`.

Cannot make `(0,1)` from `(1,0)`.

Now add

```
v3 = (1, 1)
```

Notice

```
v3 = v1 + v2
```

So `v3` not new information.

Dependent.

---

AI example

Suppose predicting house price.

Features:

```
Feature 1: Height in meters
Feature 2: Weight in kg
Feature 3: Age
```

Good.

Each tells something different.

Now imagine

```
Feature 1: Height in cm
Feature 2: Height in meters
```

But

```
Height(cm) = 100 × Height(m)
```

Same information.

One feature completely determined by other.

Model asks:

"Should weight go to Feature 1 or Feature 2?"

No unique answer.

Maybe

```
Weight1 = 5
Weight2 = -4
```

or

```
Weight1 = 100
Weight2 = -99
```

Both produce nearly same prediction.

Weights become unstable.

Tiny data change → huge weight change.

This called **multicollinearity**.

---

Easy rule:

* Independent = each feature adds new information.
* Dependent = one feature can be built from others.
* AI wants independent features because model learns stable, meaningful weights.


### The Concept
- Vectors Are Points (and Directions)
- A vector is just a list of numbers. But those numbers mean something -- they're   
coordinates in space.



## 04 Calculus for Machine Learning - Concept Is not clear


#### The gradient: vector of all partial derivatives
The gradient collects every partial derivative into one vector. For a function f(x, y, z), the gradient is:

```
grad f = [ df/dx, df/dy, df/dz ]

```



## 05 Chain Rule & Automatic Differentiation - Concepts Need To Be Much More Clear
